src/dataset.jl

```diff
@@ -172,13 +172,111 @@ end
+* `strict` (Bool): Set this to `true` if the written dataset should be a 1:1 copy of the source data, default is `false`,
+which allows the driver to adapt if necessary.
+    
+### Returns
+`nothing`
 """
 function write(
     dataset::AbstractDataset,
     filename::AbstractString;
     kwargs...,
 )::Nothing
-    destroy(unsafe_copy(dataset, filename = filename; kwargs...))
+    if nraster(dataset) > 0 && nlayer(dataset) > 0
+        drivername = :driver ∈ keys(kwargs) ? shortname(kwargs[:driver]) : shortname(getdriver(dataset))
+        error("Writing datasets with raster and vector data is not supported when using driver $drivername. 
+        Please file an issue at https://github.com/yeesian/ArchGDAL.jl/issues 
+        including following dataset information: \n\n$dataset")
+    elseif nraster(dataset) > 0
+        destroy(unsafe_copy(dataset, filename = filename; kwargs...))
+    elseif nlayer(dataset) > 0
+        writelayers(dataset, filename; kwargs...)
+    end
+    return nothing
+end
+
+"""
+    writelayers(ds, filename; driver, options=[""], layer_options=[""], chunksize=20000)
+
+Writes the vector dataset to the designated filename. The options are passed to the newly created dataset and
+have to be given as a list of strings in KEY=VALUE format. The chunksize controls the number of features written 
+in each database transaction, e.g. for SQLite. This function can also be used to copy datasets on disk.
+
+Currently working drivers: FlatGeobuf, GeoJSON, GeoJSONSeq, GML, GPKG, JML, KML, MapML, ESRI Shapefile, SQLite
+
+### Parameters
+*`dataset`  The source dataset
+*`filename` The file name to write to
+
+### Keyword arguments
+`driver`           The driver to use, you have to manually select the right driver for the file extension you wish
+`options`   A vector of strings containing KEY=VALUE pairs for driver-specific creation options
+`layer_options`    Driver specific options for layer creation
+`chunksize`        Number of features to write in one database transaction. Neglected when `use_gdal_copy` is true.
+`use_gdal_copy`    Set this to true if you encounter errors when writing. Set to false for higher speed.
+
+### Returns
+nothing
+"""
+function writelayers(dataset::AbstractDataset,
+                     filename;
+                     driver=getdriver(dataset),
+                     options::Vector{String}=[""], 
+                     layer_options::Vector{String}=[""],
+                     chunksize=20000,
+                     use_gdal_copy=false)
+    create(filename; driver=driver, options=options) do target
+        for layeridx in 0:nlayer(dataset)-1
+            sourcelayer = getlayer(dataset, layeridx)
+            sourcelayerdef = layerdefn(sourcelayer)
+            
+            if shortname(driver) == "GPKG"
+                if !any(Base.contains.(layer_options, "GEOMETRY_NAME"))
+                    if ngeom(sourcelayer) > 0  # for GPKG there can be only one geometry column / layer
```

Comment:
If this is true, can we validate that there can be "only one geometry column" and "only one layer" (they are not the same thing) for GPKG?

I.e. we need two checks,

```julia
if shortname(driver) == "GPKG"
    if layeridx > 0
        error("expected nlayer=1 for GPKG driver, got nlayer=$(nlayer(dataset)).")
    end
    ...
```
and
```julia
if ngeom(sourcelayer) > 0
    if ngeom(sourcelayer) != 1
        error("expected ngeom=1 for GPKG driver, got ngeom=$(ngeom(sourcelayer)).")
    end
    ...
```

Reply:
It seems to me that [geopackage can have multiple layers](https://gis.stackexchange.com/questions/287997/counting-the-number-of-layers-in-a-geopackage), in which case I think `push!(layer_options, "GEOMETRY_NAME=$geometry_column_name")` further beyond might introduce multiple entries of `GEOMETRY_NAME=`, and I haven't seen corresponding test cases for how it will behave in that setting.

Reply:
Ok I have to sort out the multiple layer case. But I think the "GEOMETRY_NAME" option won't be multiplied as of now, as I check for its presence and only push if it's not there.